### PR TITLE
Linter native primitives only, tested for Defined functions for now

### PR DESCRIPTION
**Pull request by StefanoDalMas**
## Description 

This is my first implementation of the linter described [here.](https://github.com/MystenLabs/sui/issues/7197)
The linter in question is:
To check: native functions should only accept primitive types or references
Why: if you pass a Move struct into Rust code, there are no ability protections and it is easy to accidentally violate ability constraints. Also, it is very cumbersome to work with Move structs inside Rust.
Severity: High. Blast radius of a native function is much higher when you break this rule

## Test Plan 

I tried this linter on defined functions(because I can't create native ones), but by commenting the Defined and uncommenting the Native piece of code it should check for Native ones.

---

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                    let uid_msg = "Only native types or references are allowed";
+                    let mut d = diag!(NATIVE_PRIMITIVE_DIAG, (sloc, msg),);
+                    d.add_note(uid_msg);
+                    env.add_diag(d);
+
+                    println!("Object in definition, time to shoot");
+                }
+            }
+        }
+        T::FunctionBody_::Native => {
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:71]: These should be flipped, as in we only want to perform this check on native functions

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            env.pop_warning_filter_scope();
+        }
+    }
+}
+
+fn check_native(env: &mut CompilationEnv, fname: Symbol, fun: &T::Function, sloc: Loc) {
+    match fun.body.value {
+        T::FunctionBody_::Defined(_) => {
+            let parameters = &fun.signature.parameters;
+            for (_var, data) in parameters.iter() {
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:56]: nit, probably something like `type_` instead of `data`

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                if !is_native_or_ref(data.value.clone()) {
+                    //This is a non-native type, so we prepare the message
+                    let c = _var.value.name;
+                    let msg =
+                        format!("The parameter '{c}' of '{fname}' is not native or a reference");
+                    let uid_msg = "Only native types or references are allowed";
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:62]: We probably want to make this a secondary label and explain that it is recommended when implementing native functions to only deal in primitives

File: crates/sui-move-build/src/linters/native_primitives_only.rs (hunk 2 above)
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:56]:
```suggestion
            for (var, type_) in parameters {
```

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+fn check_native(env: &mut CompilationEnv, fname: Symbol, fun: &T::Function, sloc: Loc) {
+    match fun.body.value {
+        T::FunctionBody_::Defined(_) => {
+            let parameters = &fun.signature.parameters;
+            for (_var, data) in parameters.iter() {
+                if !is_native_or_ref(data.value.clone()) {
+                    //This is a non-native type, so we prepare the message
+                    let c = _var.value.name;
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:59]: The leading underscore in `_var` would imply it is unused, which isn't the case here :)

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+                    let msg =
+                        format!("The parameter '{c}' of '{fname}' is not native or a reference");
+                    let uid_msg = "Only native types or references are allowed";
+                    let mut d = diag!(NATIVE_PRIMITIVE_DIAG, (sloc, msg),);
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:63]:
```suggestion
                    let mut d = diag!(NATIVE_PRIMITIVE_DIAG, (sloc, msg));
```

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+pub struct NativePrimitivesOnlyVisitor;
+
+impl TypingVisitor for NativePrimitivesOnlyVisitor {
+    fn visit(
+        &mut self,
+        env: &mut CompilationEnv,
+        _program_info: &ProgramInfo,
+        program: &mut T::Program,
+    ) {
+        for (_, _, mdef) in program.modules.iter() {
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:42]:
```suggestion
        for (_, _, mdef) in &program.modules {
```

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            env.add_warning_filter_scope(mdef.warning_filter.clone());
+            mdef.functions
+                .iter()
+                .for_each(|(sloc, fname, fun)| check_native(env, *fname, fun, sloc));
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:46]: A `for` loop here might be more canonical, but this is fine?

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    fn visit(
+        &mut self,
+        env: &mut CompilationEnv,
+        _program_info: &ProgramInfo,
+        program: &mut T::Program,
+    ) {
+        for (_, _, mdef) in program.modules.iter() {
+            env.add_warning_filter_scope(mdef.warning_filter.clone());
+            mdef.functions
+                .iter()
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:45]: `
```suggestion
            mdef.functions
                .key_cloned_iter()
```
Using `key_cloned_iter` would be more in line with the rest of the compiler

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+            //         d.add_note(uid_msg);
+            //         env.add_diag(d);
+            //
+            //         println!("Object in definition, time to shoot");
+            //     }
+            // }
+        }
+    }
+}
+fn is_native_or_ref(element: N::Type_) -> bool {
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:91]: Changes this to a reference so you do not have to clone

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+                return false;
+            }
+        },
+        N::Type_::Ref(_is_mutable, _referenced_element) => {
+            //I'm ok, it's a reference, I can't have && in move
+            return true;
+            // let el = referenced_element.as_ref().value.clone();
+            // is_native_or_ref(el);
+        }
+        _ => {}
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:111]: Nit, we should exhaustively match here

Comment (tnowacki) [reply crates/sui-move-build/src/linters/native_primitives_only.rs:111]: If you change this branch to `false` you do not need the `return`s above :)

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+                //Flash out that a struct should not be here!
+                return false;
+            }
+            _ => {
+                return false;
+            }
+        },
+        N::Type_::Ref(_is_mutable, _referenced_element) => {
+            //I'm ok, it's a reference, I can't have && in move
+            return true;
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:107]: We need to recurse he to make sure we have a reference to a primitive

File: crates/sui-move-build/tests/linter/native_primitives_only.move [hunk 12]
```diff
@@ -0,0 +1,22 @@
+
+#[allow(unused_variable)]
+module 0x42::test1{
+    //has copy needed because it's a vector
+
+    struct S1 has copy{}
+    
+    struct Coolstruct has copy,drop{
```
Comment (tnowacki) [inline crates/sui-move-build/tests/linter/native_primitives_only.move:8]:
```suggestion
    struct Coolstruct has copy, drop {
```

File: crates/sui-move-build/tests/linter/native_primitives_only.move [hunk 13]
```diff
@@ -0,0 +1,22 @@
+
+#[allow(unused_variable)]
+module 0x42::test1{
+    //has copy needed because it's a vector
```
Comment (tnowacki) [inline crates/sui-move-build/tests/linter/native_primitives_only.move:4]: Not sure what this comment is for?

File: crates/sui-move-build/tests/linter/native_primitives_only.move [hunk 14]
```diff
@@ -0,0 +1,22 @@
+
+#[allow(unused_variable)]
```
Comment (tnowacki) [inline crates/sui-move-build/tests/linter/native_primitives_only.move:2]: nit, remove and use leading `_` if needed

File: crates/sui-move-build/tests/linter/native_primitives_only.move [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+module 0x42::test1{
```
Comment (tnowacki) [inline crates/sui-move-build/tests/linter/native_primitives_only.move:3]:
```suggestion
module 0x42::test1 {
```

File: crates/sui-move-build/tests/linter/native_primitives_only.move [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+    struct Coolstruct has copy,drop{
+        a: bool,
+        b: u64,
+    }
+
+    #[allow(unused_function)]
+    fun returns_something(a:bool,b:u64,c:Coolstruct,d:&Coolstruct) : (bool,u64){
+        let x = b;
+        (a,x)
+    }
```
Comment (tnowacki) [inline crates/sui-move-build/tests/linter/native_primitives_only.move:17]:
```suggestion
    #[allow(unused_function)]
    native returns_something(a: bool, b: u64, c: Coolstruct, d: &Coolstruct): (bool, u64);
```

File: crates/sui-move-build/tests/linter/native_primitives_only.move [hunk 17]
```diff
@@ -0,0 +1,22 @@
+
```
Comment (tnowacki) [inline crates/sui-move-build/tests/linter/native_primitives_only.move:1]: We will also want to comment these and add a copyright. 
And we should probably split this into positive/negative tests

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+            env.add_warning_filter_scope(mdef.warning_filter.clone());
+            mdef.functions
+                .key_cloned_iter()
+                .for_each(|(fname, fdef)| check_native(env, fname.value(), fdef, fname.loc()));
+            env.pop_warning_filter_scope();
+        }
+    }
+}
+
+fn check_native(env: &mut CompilationEnv, fname: Symbol, fun: &T::Function, sloc: Loc) {
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:46]:
```suggestion
fn check_native(env: &mut CompilationEnv, fname: FunctionName, fun: &T::Function) {
```

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+        &mut self,
+        env: &mut CompilationEnv,
+        _program_info: &ProgramInfo,
+        program: &mut T::Program,
+    ) {
+        for (_, _, mdef) in &program.modules {
+            env.add_warning_filter_scope(mdef.warning_filter.clone());
+            mdef.functions
+                .key_cloned_iter()
+                .for_each(|(fname, fdef)| check_native(env, fname.value(), fdef, fname.loc()));
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:40]:
```suggestion
            for (fname, fdef) in mdef.functions.key_cloned_iter() {
                check_native(env, fname.value(), fdef, fname.loc()));
            }
```

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+
+pub struct NativePrimitivesOnlyVisitor;
+
+impl TypingVisitor for NativePrimitivesOnlyVisitor {
+    fn visit(
+        &mut self,
+        env: &mut CompilationEnv,
+        _program_info: &ProgramInfo,
+        program: &mut T::Program,
+    ) {
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:35]: You can replace this with `TypingVisitorConstructor` and `TypingVisitorContext` now that it has landed. It should reduce the amount of boilerplate you need to do

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+        }
+    }
+}
+
+fn check_native(env: &mut CompilationEnv, fname: Symbol, fun: &T::Function, sloc: Loc) {
+    match fun.body.value {
+        T::FunctionBody_::Defined(_) => {}
+        T::FunctionBody_::Native => {
+            let parameters = &fun.signature.parameters;
+            for (var, type_) in parameters.iter() {
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:51]:
```suggestion
            for (var, type_) in &fun.signature.parameters {
```

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+                if !is_native_or_ref(&type_.value) {
+                    //This is a non-native type, so we prepare the message
+                    let c = var.value.name;
+                    let msg =
+                        format!("The parameter '{c}' of '{fname}' is not native or a reference");
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:56]:
```suggestion
                    let c = var.value.name;
                    let msg =
                        format!("The parameter '{var}' is not a primitive or a reference to a primitive");
```

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+                    let uid_msg = "- It is recommended when implementing native functions to only deal with primitives";
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:57]:
```suggestion
                    let fun_msg = "When possible, native function parameters should have primitive types";
```

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+                    let d = diag!(NATIVE_PRIMITIVE_DIAG, (var.loc, msg), (sloc, uid_msg));
+                    env.add_diag(d);
+
+                    // println!("Object in definition, time to shoot");
+                }
+            }
+        }
+    }
+}
+fn is_native_or_ref(element: &N::Type_) -> bool {
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:67]:
```suggestion

fn is_native_or_ref(sp!(_, ty_): &N::Type) -> bool {
```

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+            }
+            N::TypeName_::Multiple(_) => {
+                return false;
+            }
+        },
+        N::Type_::Ref(_is_mutable, referenced_element) => {
+            //I have to check if it's a reference to a native type or not
+            let el = &referenced_element.value;
+            is_native_or_ref(el)
+        }
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:85]:
```suggestion
        N::Type_::Ref(_is_mutable, inner) => is_native_or_ref(&inner)
```

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+                return true;
+            }
+            N::TypeName_::ModuleType(_mident, _sname) => {
+                //Flash out that a struct should not be here!
+                return false;
+            }
+            N::TypeName_::Multiple(_) => {
+                return false;
+            }
+        },
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:80]:
```suggestion
        N::Type_::Apply(_, typename, ty_args) => match &typename.value {
            N::TypeName_::ModuleType(_, _) | N::TypeName_::Multiple(_) => false,
            N::TypeName_::Builtin(_) => ty_args.iter().all(|ty| is_native_or_ref),
        },
```

File: crates/sui-move-build/src/linters/native_primitives_only.rs [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+        }
+        N::Type_::Anything => {
+            return false;
+        }
+        N::Type_::UnresolvedError => {
+            return false;
+        }
+        N::Type_::Param(_) => {
+            return false;
+        }
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/native_primitives_only.rs:100]:
```suggestion
        N::Type_::Unit
        | N::Type_::Var(_) 
        | N::Type_::Anything 
        | N::Type_::UnresolvedError
        | N::Type_::Param(_) => false,
```

File: crates/sui-move-build/tests/linter/native_primitives_only.move (hunk 17 above)
Comment (tnowacki) [reply crates/sui-move-build/tests/linter/native_primitives_only.move:3]: We still need a comment about what is being tested, and about we should still split this into positive/negative tests

**Review by tnowacki — COMMENTED** [review]
Thanks for the progress! @awelc, would you be able to help switch to the new typing visitor?
